src/server.rs

```diff
@@ -29,35 +29,63 @@ pub fn listen<S, Req, Resp, E>(addr: SocketAddr, new_service: S) -> ListenFuture
           E: Serialize + 'static
 {
     let (tx, rx) = futures::oneshot();
-    REMOTE.spawn(move |handle| Ok(tx.complete(listen_with(addr, new_service, handle))));
+    REMOTE.spawn(move |handle| Ok(tx.complete(listen_with(addr, new_service, handle.clone()))));
     ListenFuture { inner: rx }
 }
 
 /// Spawns a service that binds to the given address using the given handle.
 pub fn listen_with<S, Req, Resp, E>(addr: SocketAddr,
                                     new_service: S,
-                                    handle: &Handle)
-                                    -> io::Result<ServerHandle>
+                                    handle: Handle)
+                                    -> io::Result<SocketAddr>
     where S: NewService<Request = Result<Req, DeserializeError>,
                         Response = Response<Resp, E>,
                         Error = io::Error> + Send + 'static,
           Req: Deserialize + 'static,
           Resp: Serialize + 'static,
           E: Serialize + 'static
 {
-    server::listen(handle, addr, move |stream| {
-        Ok(multiplex::EasyServer::new(new_service.new_service()?, Framed::new(stream))
-               .map_err(|()| panic!("What do we do here")))
+    let listener = listener(&addr, &handle)?;
+    let addr = listener.local_addr()?;
+
+    let handle2 = handle.clone();
+    let server = listener.incoming().for_each(move |(socket, _)| {
+        Proto::new().bind_server(&handle2, socket, new_service.new_service()?);
+
+        Ok(())
+    }).map_err(|e| error!("While processing incoming connections: {}", e));
```

Comment:
I've never triggered it in testing but it can definitely happen. I guess if there's a problem accepting a request